extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/devconsole/ArcDevConsoleProcessor.java

```diff
@@ -183,38 +200,52 @@ void addNodesDependencies(BeanInfo root, Set<DevBeanInfo> nodes, Set<Link> links
         }
     }
 
-    void addNodesDependents(BeanInfo root, Set<DevBeanInfo> nodes, Set<Link> links, BeanInfo bean, Collection<BeanInfo> beans,
-            Map<BeanInfo, List<BeanInfo>> declaringToProducers, BeanResolver resolver, DevBeanInfos devBeanInfos) {
-        for (BeanInfo dependent : beans) {
-            if (!bean.equals(dependent)) {
-                for (InjectionPointInfo injectionPoint : dependent.getAllInjectionPoints()) {
-                    Link link = null;
-                    if (injectionPoint.isProgrammaticLookup()) {
-                        if (resolver.matches(bean,
-                                injectionPoint.getType().asParameterizedType().arguments().get(0),
-                                injectionPoint.getRequiredQualifiers())) {
-                            link = Link.lookup(dependent.getIdentifier(), bean.getIdentifier());
-                        }
-                    } else if (bean.equals(injectionPoint.getResolvedBean())) {
-                        link = Link.dependent(root.equals(bean), dependent.getIdentifier(), bean.getIdentifier());
-                    }
-                    if (link != null) {
-                        links.add(link);
-                        if (nodes.add(devBeanInfos.getBean(dependent.getIdentifier()))) {
-                            // add transient dependents
-                            addNodesDependents(root, nodes, links, dependent, beans, declaringToProducers, resolver,
-                                    devBeanInfos);
-                        }
+    void addNodesDependents(BeanInfo root, Set<DevBeanInfo> nodes, Set<Link> links, BeanInfo bean,
+            List<InjectionPointInfo> injectionPoints, Map<BeanInfo, List<BeanInfo>> declaringToProducers, BeanResolver resolver,
+            DevBeanInfos devBeanInfos, Map<BeanInfo, List<InjectionPointInfo>> directDependents) {
+        List<InjectionPointInfo> direct = directDependents.get(bean);
+        if (direct == null) {
+            direct = new ArrayList<>();
+            for (InjectionPointInfo injectionPoint : injectionPoints) {
```

Comment:
Naive question, but this loop has to run once for each bean, while it seems it should be possible to create a map of direct dependencies (either `Map<BeanInfo, List<InhjectionPointInfo>>` or even `Map<BeanInfo, List<BeanInfo>>`) once, when building the dependency graph (that is, when the `allInjectionPoints` list is being populated). That should be quite a bit faster when the number of injections is much higher than the number of beans I'd guess?

Reply:
So in this particular case we would need an "inverse" map because we're looking for _dependents_ not dependencies. And since we cache the result I believe it's essentially the same perf as if was built beforehand, or am I missing something?

Reply:
I think @Ladicek is basically suggesting to move this logic to L#144 where you iterate over all beans and their IPs so that you could build the same information from just a single iteration.

Reply:
@manovotn is right on the money. Single iteration should be enough and should provide some additional speedup (I _think_ :-) )

Reply:
Looking more in the code, it seems like that would work well in resolving direct dependencies while saving some iterations (and potentially a lot of them in cases with lots of IPs for each bean), so +1
You'd still need to so the transient dependency resolution and producers later on once you have complete information on direct dependencies, but unless I am missing something, that seems doable.

Reply:
@mkouba something like this should do the trick - https://github.com/manovotn/quarkus/commit/3c4067977e5178ece195b779d7c96e79e61cb18e
Feel free to use that commit however you wish (warning: didn't run more than a quickstart sanity check with it)

Reply:
Big props, I considered refactoring the code myself, but was too lazy :laughing:

Reply:
Hm, I get your point. But this would not help for programmatic lookup injection points that would require an additional iteration over all beans to find all eligible candidates. In other words, it would probably help a little when the number of injections is much higher than the number of beans but could make it worse when it's the other way around.

Reply:
I'm not sure I follow -- the way I understand it, @manovotn's change leads to strictly less work than before when building the dependents map, and all the other work remains unchanged?

Reply:
After some additional discussion over Zulip, I'm retracting this comment. Thanks for your patience though :-)